### PR TITLE
build: TestProtoMarshal correctly avoids Marshal methods

### PR DESCRIPTION
**Pull request by tamird**
Also use `git grep` exclusions rather than filtering its results.

### CODE REVIEW

**Comment by benesch** [comment]
This is on my radar! Just still catching up from vacation.

**Comment by tamird** [comment]
Rebased to fix a merge conflict.

**Comment by nvanbenschoten** [comment]
:lgtm:

@CLAassistant where you at?

---

Reviewed 20 of 20 files at r1.
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[build/style_test.go, line 139 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18496#-KuSbJWc9Fp3NCPphJuy:-KuSbJWc9Fp3NCPphJuz:bs474mn) ([raw file](https://github.com/cockroachdb/cockroach/blob/667ce2d2138bcc0c334d4bc99953b414c280f438/build/style_test.go#L139)):*
> ```Go
> 				re: `\bos\.(Getenv|LookupEnv)\(`,
> 				excludes: []string{
> 					":!acceptance",
> ```

Where does a layman find documentation on this `:!` exclusion syntax?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/18496#-:-KuScng906XHgkeEU5YK:b-8al1yx)*

**Comment by tamird** [comment]
Good question, I think I got it from some SO thread. Thanks!

**Comment by tamird** [comment]
Rebased to avoid skew; looks like acceptance hung, and my teamcity privileges have been revoked :(

Can someone rebuild for me?

**Comment by benesch** [comment]
Queued you a new build!

---

Review status: 19 of 20 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[build/style_test.go, line 139 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18496#-KuSbJWc9Fp3NCPphJuy:-KuUCrJX5demsx5pCyXr:b-fztly7) ([raw file](https://github.com/cockroachdb/cockroach/blob/667ce2d2138bcc0c334d4bc99953b414c280f438/build/style_test.go#L139)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Where does a layman find documentation on this `:!` exclusion syntax?
</blockquote></details>

https://git-scm.com/docs/gitglossary/2.13.1#gitglossary-aiddefpathspecapathspec

(Git pathspecs are _super_ powerful.)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/18496#-:-KuUCYifkqzT-cMsZnnT:b2qy5dn)*

File: build/style_test.go [hunk 1]
```diff
@@ -129,38 +129,72 @@ func TestStyle(t *testing.T) {
...
+		for _, tc := range []struct {
+			re       string
+			excludes []string
+		}{
+			{re: `\bos\.(Getenv|LookupEnv)\("COCKROACH`},
+			{
+				re: `\bos\.(Getenv|LookupEnv)\(`,
+				excludes: []string{
+					":!acceptance",
+					":!build/style_test.go",
```
Comment (benesch) [inline build/style_test.go]: This isn't in `pkg`, is it? Why would it get picked up?

File: build/style_test.go [hunk 2]
```diff
@@ -294,7 +355,17 @@ func TestStyle(t *testing.T) {
...
+			pkg.Dir,
+			"git",
+			"grep",
+			"-nE",
+			`\.Marshal\(`,
+			"--",
+			"*.go",
+			":!util/protoutil/marshal.go",
+			":!settings/settings_test.go",
+		)
```
Comment (benesch) [inline build/style_test.go]: This pattern really seems to deserve a `grepExcluding(`regexpattern`, "file", "file2", ...)` function, but this is an improvement as is so no need to hold up this PR.

File: build/style_test.go [hunk 3]
```diff
@@ -649,6 +727,9 @@ func TestStyle(t *testing.T) {
 	// TODO(tamird): replace this with errcheck.NewChecker() when
 	// https://github.com/dominikh/go-tools/issues/57 is fixed.
 	t.Run("TestErrCheck", func(t *testing.T) {
+		if testing.Short() {
+			t.Skip("short flag")
+		}
```
Comment (benesch) [inline build/style_test.go]: Any particular reason you're sidecar-ing this bit?

**Review by benesch — COMMENTED** [review]
(Sorry for the post-hoc review. If you don't mind responding to my comments with your thoughts I'm happy to pick off any changes myself.)

File: build/style_test.go (hunk 1 above)
Comment (tamird) [reply build/style_test.go]: I think `git grep` doesn't care about the directory you're in, or at least it doesn't when you specify `'*.go'`.

File: build/style_test.go (hunk 2 above)
Comment (tamird) [reply build/style_test.go]: Eh, I'm not sure I agree that there's much benefit to the indirection. Feel free to follow up if you feel strongly enough.

**Pull request merged**